### DISCUSSION TITLE
How to run ZEN-garden on the EULER cluster

### DISCUSSION THREAD

**Discussion by maximilian-hoffmann** [ZEN-universe/ZEN-garden#169] [Installation & Setup]
**+++ THIS IS OUTDATED, PLEASE FOLLOW DISCUSSION #186 +++**

Prerequisites:
1. Ask Andrej to request access to the EULER cluster for you (Access may take a day).
2. Download mobaxterm.
3. Access the cluster via mobaxterm using "Session" (Top right) and enter "euler.ethz.ch" in the pop-up window next to "remote host" and click "ok".
4. Log in by entering you mail identifier, e.g. "mmuster" and your correspoding LDAP password.
5. Navigate to your "mmuster" folder.

Installation:
1. Clone the github repository using `git clone https://github.com/RRE-ETH/ZEN-garden.git`.
2. Create access token on github by clicking on your profile thumbnail->settings->developer settings->personal access tokens->generate new token and then setting the name and the required ticks. In the command on mobaxterm, you're asked for your github name and the password, which is the long ssh-link from the generated access token on github.
3. Create and activate a virtual python environment called zen-garden:
`$ module load gcc/6.3.0 python/3.8.5`
`$ python -m venv --system-site-packages zen-garden`
`$ source zen-garden/bin/activate`
For deactivation:
`$ deactivate zen-garden`
5. If you have a runnable version of zen-garden locally, activate the corresponding environment in the windows powershell locally and export all installed packages to a requirements.txt using `pip list --format=freeze > requirements.txt`.
6. Copy the requirements.txt to your "mmuster" folder, to which you already cloned the zen-garden folder.
7. As some packages are causing problems on linux on the cluster, **comment GDAL, pywin32 and pywinpty out** in the requirements.txt using "#". In the mobaxterm terminal, run `pip install -r requirements.txt` (this is the basic workaround for **not** using conda on the cluster, which is recommended not to use.) Find the requirements.txt here: [requirements.txt](https://github.com/RRE-ETH/ZEN-garden/files/9022549/requirements.txt)
8. For some troubleshooting for eccodes issue, run:
`$ pip install ecmwflibs`
`$ pip install eccodes==1.3.1`
`$ pip install cfgrib`
9. Important: In order to make pyomo find the gurobi folders, you need to install gurobipy 9.5.1 and load the module. Therefore run in the terminal:
`pip install gurobipy==9.5.1`
`module load gurobi/9.5.1`
10. Access and run the compile.py using e.g.:
`$ cd ZEN-garden`
`git chekout <your_branch>`
`$ bsub -n 42 -R "span[hosts=1]" -R "model==EPYC_7H12" -R "rusage[mem=3200]" -W "24:00" python compile.py`
[requirements.txt](https://github.com/RRE-ETH/ZEN-garden/files/9022549/requirements.txt)
with slurm: `$ sbatch -n 42 --ntasks-per-node=1 --constraint=EPYC_7H12 --mem-per-cpu=3200 -t 02:00 --wrap="python -m zen_garden --dataset <dataset name>"`
[LSF to Slurm](https://scicomp.ethz.ch/wiki/LSF_to_Slurm_quick_reference)

If you are done, don't forget to exit mobaxterm regularly:
`$ deactivate` (the zen-garden environment)
`$ exit`

Before the next start, always load the python and gurobi modules first:
`$ module load gcc/6.3.0 python/3.8.5`
`$ module load gurobi/9.5.1`

Use the cluster, leverage performance.
🥚 ▶️ 🦖

**Comment by maximilian-hoffmann** [comment]
[RREGuide_Euler instructions.docx](https://github.com/RRE-ETH/ZEN-garden/files/9012350/RREGuide_Euler.instructions.docx)

**Comment by jacob-mannhardt** [comment]
https://scicomp.ethz.ch/wiki/How_to_access_the_cluster_with_MobaXterm

**Reply by jacob-mannhardt** [reply]
Prerequisites:

1. Ask Andrej to request access to the EULER cluster for you (Access may take a day).
2. Download mobaxterm.
3. Access the cluster via mobaxterm using "Session" (Top left), "SSH session" and enter "euler.ethz.ch" in the pop-up window next to "remote host" and click "ok".
4. Log in by entering you mail identifier, e.g. "mmuster" and your correspoding LDAP password.
5. Navigate to your "mmuster" folder.

Installation:

1. Create access token on github by clicking on your profile thumbnail->settings->developer settings->personal access tokens->generate new token and then setting the name _**and the required ticks**_. In the command on mobaxterm, you're asked for your github name and the password, which is the long ssh-link from the generated access token on github.
2. Clone the github repository using `git clone https://github.com/RRE-ETH/ZEN-garden.git`. If you want to clone a specific branch `git clone -b branchname https://github.com/RRE-ETH/ZEN-garden.git`
4. Follow Janis description here #186
